Feat: http request retry logic :repeat:

mikonieminen opened the pull request:
Logic for retrying HTTP request in some conditions like when the gateway returns 5xx.

mikonieminen committed:
Implement retry logic component

mikonieminen committed:
Add FIXME comment

mikonieminen committed:
Use single abstraction for HTTP client

mikonieminen committed:
Improve HTTP client abstraction

Minor improvement in HTTP client abstraction to allow mock client to
wrap around implementation specific error type.

For now this is needed, because this abstraction leaks through
the actual implementation and error types were limiting usefulness
of this abtraction.

mikonieminen committed:
Add retry logic for Arweave network info query

mikonieminen commented on src/cron/arweave.rs:
```diff
@@ -159,25 +184,53 @@ impl Arweave {
-        } else {
-            Err(res.error_for_status().err().unwrap()) // FIXME: do not unwrap
+        let client = ctx.get_client();
+        let res = retry::<tokio::runtime::Handle, _>()
+            .max_retries(3)
+            .run_with_context(&(client, uri), |(client, uri)| async move {
+                let req: http::Request<String> = http::request::Builder::new()
+                    .method(http::Method::GET)
+                    .uri(uri)
+                    .body("".to_string())
+                    .unwrap();
+                let req: reqwest::Request = reqwest::Request::try_from(req).unwrap();
+                let res = client.execute(req).await;
+                match res {
+                    Ok(res) => {
+                        if let Some(retry_after) = res.headers().get(http::header::RETRY_AFTER) {
+                            let retry_delay = match retry_after.to_str().unwrap().parse().unwrap() {
+                                RetryAfter::Duration(duration_seconds) => {
+                                    Duration::seconds(duration_seconds).to_std().unwrap()
+                                }
+                                RetryAfter::Timestamp(timestamp) => timestamp
+                                    .signed_duration_since(Utc::now())
+                                    .to_std()
+                                    .unwrap(),
+                            };
+                            RetryControl::Retry(Ok(res), Some(retry_delay))
+                        } else if res.status().is_server_error() {
+                            RetryControl::Retry(Ok(res), None)
+                        } else {
+                            RetryControl::Success(Ok(res))
+                        }
+                    }
+                    ret @ Err(_) => RetryControl::Fail(ret),
+                }
+            })
+            .await?;
+
+        match res.error_for_status() {
+            Ok(res) => res.json().await.map_err(|err| err.into()),
+            Err(err) => Err(err.into()),
```
This is the main part of the change, repeating the request when receiving `Retry-After` header or server error.

mikonieminen replied:
This could be encapsulated into a function to make the code a bit more readable and to allow reusing this retry logic. I'll do that if we agree over the logic and also if we need the retry in some other parts of the code.

bluekirby1111 replied:
Looks good. It would be nice to have this retry logic, with a `max_retries` parameter, separated and reusable, for sure.

bluekirby1111 reviewed (commented):
LGTM, based on main logic commented previously.
It would be really nice to move the retry logic into a separate module, maybe on `utils`

mikonieminen committed:
Move retry logic for HTTP requests into a function

Make it easier to reuse the retry logic for failed HTTP requests by
moving the logic into a separate function. For now, this is Reqwest
implementation specific one to simplify things, but later we should try
to create generic implementation that is then implemented by each HTTP
client integration. Currently it's just a bit tricky because the
implementation requires quite complex typing around futures and
lifetimes.
